lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskHostBase.cs

```diff
@@ -0,0 +1,242 @@
+
+        /// <summary>
+        /// Task close handler. Call TaskCloseCoordinator to handle the event.
+        /// </summary>
+        /// <param name="closeEvent"></param>
+        public void OnNext(ICloseEvent closeEvent)
+        {
+            _taskCloseCoordinator.HandleEvent(closeEvent, CancellationSource);
+        }
+
+        /// <summary>
+        /// Dispose function. Dispose IGroupCommunicationsClient.
+        /// </summary>
+        public void Dispose()
+        {
+            if (Interlocked.Exchange(ref _disposed, 1) == 0)
+            {
+                _groupCommunicationsClient.Dispose();
+            }
+        }
+
+        public void OnError(Exception error)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void OnCompleted()
+        {
+            throw new NotImplementedException();
+        }
+
+        /// <summary>
+        /// Convert the exception into IMRUTaskGroupCommunicationException
+        /// </summary>
+        /// <param name="e"></param>
+        protected void HanldeCommunicationException(Exception e)
+        {
+            HanldeException(e, new IMRUTaskGroupCommunicationException(TaskManager.TaskGroupCommunicationError));
+        }
+
+        /// <summary>
+        /// Convert the exception into IMRUSystemException
+        /// </summary>
+        /// <param name="e"></param>
+        protected void HanldeSystemException(Exception e)
+        {
+            HanldeException(e, new IMRUSystemException(TaskManager.TaskSystemError));
+        }
+
+        /// <summary>
+        /// Convert the exception into IMRUTaskAppException
+        /// </summary>
+        /// <param name="e"></param>
+        protected void HanldeTaskAppException(Exception e)
+        {
+            HanldeException(e, new IMRUTaskAppException(TaskManager.TaskAppError));
+        }
+
+        /// <summary>
+        /// Log and throw target exception if cancellation token is not set
+        /// In the cancellation case, simply log and return.
+        /// </summary>
+        /// <param name="originalException"></param>
+        /// <param name="targetException"></param>
+        private void HanldeException(Exception originalException, Exception targetException)
+        {
+            Logger.Log(Level.Error,
+                "Received Exception {0} in {1} with message: {2}. The cancellation token is: {3}.",
+                originalException.GetType(),
+                TaskHostName,
+                originalException.Message,
+                CancellationSource.IsCancellationRequested);
+            if (!CancellationSource.IsCancellationRequested)
+            {
+                Logger.Log(Level.Error,
+                    "{0} is throwing {1} with cancellation token: {2}.",
+                    TaskHostName,
+                    targetException.GetType(),
+                    CancellationSource.IsCancellationRequested);
+                throw targetException;
```

Comment:
The exception we throw here doesn't have information about the original exception. So we're logging original exception here (evaluator side), but on driver side we only get exception type (exception message is fixed for each type). Should we carry some information about original exception with `IMRU*Exception` when it goes back to driver to improve debugability? We have such constructor in `IMRU*Exception` but never use it